src/gluonts/torch/distributions/implicit_quantile_network.py

```diff
@@ -0,0 +1,211 @@
+        )
+
+        self.register_buffer("integers", torch.arange(0, cos_embedding_dim))
+
+    def forward(self, tau):  # tau: [B, T]
+        cos_emb_tau = torch.cos(tau.unsqueeze(-1) * self.integers * torch.pi)
+        return self.output_layer(cos_emb_tau)
+
+
+class ImplicitQuantileModule(nn.Module):
+    r"""
+    Implicit Quantile Network from the paper ``IQN for Distributional
+     Reinforcement Learning`` (https://arxiv.org/abs/1806.06923) by
+     Dabney et al. 2018.
+    """
+
+    def __init__(
+        self,
+        in_features: int,
+        args_dim: Dict[str, int],
+        domain_map,
+        concentration1: float = 1.0,
+        concentration0: float = 1.0,
+        output_domain_map=None,
+        cos_embedding_dim: int = 64,
+    ):
+        super().__init__()
+        self.output_domain_map = output_domain_map
+        self.domain_map = domain_map
+        self.beta = Beta(
+            concentration1=concentration1, concentration0=concentration0
+        )
+
+        self.quantile_layer = QuantileLayer(
+            in_features, cos_embedding_dim=cos_embedding_dim
+        )
+        self.output_layer = nn.Sequential(
+            nn.Linear(in_features, in_features), nn.PReLU()
+        )
+
+        self.proj = nn.ModuleList(
+            [nn.Linear(in_features, dim) for dim in args_dim.values()]
+        )
+
+    def forward(self, inputs):
+        if self.training:
+            taus = self.beta.sample(sample_shape=inputs.shape[:-1]).to(
+                inputs.device
+            )
+        else:
+            taus = torch.rand(size=inputs.shape[:-1], device=inputs.device)
+
+        emb_taus = self.quantile_layer(taus)
+        emb_inputs = inputs * (1.0 + emb_taus)
+
+        emb_outputs = self.output_layer(emb_inputs)
+        outputs = [proj(emb_outputs).squeeze(-1) for proj in self.proj]
+        if self.output_domain_map is not None:
+            outputs = [self.output_domain_map(output) for output in outputs]
+        return (*self.domain_map(*outputs), taus)
+
+
+class ImplicitQuantileNetwork(Distribution):
+    r"""
+    Distribution class for the Implicit Quantile from which
+    we can sample or calculate the quantile loss.
+
+    Parameters
+    ----------
+    outputs
+        Outputs from the Implicit Quantile Network.
+    taus
+        Tensor random numbers from the Beta or Uniform distribution for the
+        corresponding outputs.
+    """
+
+    arg_constraints = {}
+
+    def __init__(self, outputs, taus, validate_args=None):
+        self.taus = taus
+        self.outputs = outputs
+
+        super().__init__(
+            batch_shape=outputs.shape, validate_args=validate_args
+        )
+
+    @torch.no_grad()
+    def sample(self, sample_shape=torch.Size()):
+        return self.outputs
+
+    def quantile_loss(self, value) -> torch.Tensor:
+        return torch.abs(
+            (self.outputs - value)
+            * ((value <= self.outputs).float() - self.taus)
+        )
```

Comment:
Here the `abs` is not needed: when the residual is negative, the coefficient is `-taus`, making the result positive

Reply:
ah yes let me fix it! thanks!

Reply:
As a one liner, I think the most intuitive way to write this is 

```python
return (self.taus - (self.outputs < value)) * (self.outputs - value)
```

which reads “penalize by tau for over-predicting, and by 1-tau for under-predicting”